src/main/java/ecs/GameObject.java

```diff
@@ -1,273 +1,258 @@
-     */
-    public GameObject addComponent(Component c) {
-        if (this.components.stream().anyMatch(component -> c.isConflictingWith(component.getClass()) || component.isConflictingWith(c.getClass())))
-            throw new IllegalComponentStateException("Component " + c.getClass() + " is conflicting with existing one");
-        this.components.add(c);
-        if (c instanceof LocationSensitive)
-            this.transformSensitives.add((LocationSensitive) c);
-        c.gameObject = this;
-        //TODO check if this is necessary
-        if (getParentScene() != null) {
-            if (getParentScene().isActive()) {
-                c.start();
-                getParentScene().addToRenderers(this);
-            }
-        }
-        //update collision maps in scene
-        parentScene.updateGameObject(this, true);
-        return this;
-    }
-
-    /**
-     * @return List of Components in GameObject
-     */
-    public List<Component> getComponents() {
-        return components;
-    }
-}
+package ecs;
+
+import org.joml.Vector2f;
+import org.joml.Vector3f;
+import physics.LocationSensitive;
+import scene.Scene;
+import util.Engine;
+import util.OrderPreservingList;
+import java.awt.*;
+import java.util.Collection;
+import java.util.LinkedList;
+import java.util.List;
+
+/**
+ * A GameObject is the root of the Entity Component system used to store all entities in Azurite games.
+ * Each GameObject can contain any of a number of available components including spriteRenderers and lights.
+ * By default, each GameObject contains a Transform, which holds the X and Y position, and width and height of the object in pixels.
+ *
+ * @author Asher Haun
+ * @author Gabe
+ * @author Juyas
+ */
+
+public class GameObject {
+
+    public static final String DEFAULT_GAMEOBJECT_NAME = "Default GameObject Name";
+    public static final String EMPTY_GAMEOBJECT_NAME = "Empty GameObject";
+    public static final int DEFAULT_Z_INDEX = 0;
+    private static long internalCounter = 0;
+    private final long objId = internalCounter++;
+    private final OrderPreservingList<Component> components;
+    private final Scene parentScene;
+    private final Collection<LocationSensitive> transformSensitives;
+    private final String name;
+    private final Vector3f locationData;
+    private final Vector3f locationBuffer;
+    private int zIndex;
+
+    /**
+     * Creates a new GameObject.
+     *
+     * @param name
+     * @param componentList
+     * @param locationData
+     * @param zIndex
+     */
+    public GameObject(String name, List<Component> componentList, Vector3f locationData, int zIndex) {
+        this.name = name;
+        this.components = new OrderPreservingList<>(componentList);
+        this.locationData = locationData;
+        this.locationBuffer = new Vector3f();
+        this.zIndex = zIndex;
+        this.parentScene = Engine.window().currentScene();
+        Engine.window().currentScene().addGameObjectToScene(this);

```

Comment:
This is a hidden dependency we removed ealier... I am not sure, if we should do it like that.

Reply:
I am not sure what you mean by hidden, it is a publicly accessible method.  
Personally I dislike having to pass the Scene, but we could compromise with an additional constructor.

Reply:
we already did compromise on that in discord, so pls add it ;)

Reply:
This concern is now resolved.